### DISCUSSION TITLE
Integrating Supabase Authentication with Serverpod for Managing Companies, Branches, and Users

### DISCUSSION THREAD

**Discussion by iampopal** [serverpod/serverpod#2825] [Q&A]
Description:

We are developing an accounting system with a structure that includes:

1) **Companies**: For example, BankA.
2) **Branches**: Each company has multiple branches with unique names and locations.
3) **Users**: Each branch has its own set of users, each assigned specific roles.
Our goal is to manage authentication and user roles via Supabase while handling transactional data storage in Serverpod. Here’s the breakdown of our plan:

**Supabase:**

Used for managing companies, branches, users, and role-based access control (RBAC). Authentication tokens and roles will be defined here.
Serverpod:

This will handle all transactional data related to users, companies, and branches. Each transaction will include `companyId, branchId, and userId`.

**User Management in Serverpod:**

We plan to create a user in Serverpod that holds both a unique Serverpod userId and the corresponding Supabase userId. This allows us to manage data between both platforms efficiently.
Our questions and clarifications are as follows:

**1) Synchronization:** Is there a recommended best practice for synchronizing user data between Supabase and Serverpod, especially when handling user authentication tokens across systems?

**2) Token Validation:** Should we implement periodic token validation within Serverpod for enhanced security? Are there specific hooks in Serverpod that can assist with validating Supabase tokens before processing transactions?

**3) Data Integrity:** Any advice on maintaining strong data integrity in Serverpod for transactions involving multiple companies, branches, and users?

**4) Scalability:** What are the recommended patterns in Serverpod for scaling systems with potentially high transactional throughput?

**5) Multilple Database:** What is the best practice to use multiple database each database for each user in future we might plan this too. 

We would love to hear your thoughts and suggestions on this approach.

**Comment by iampopal** [comment]
Anyone, any thoughts?

**Comment by SandPod** [comment]
Hey @iampopal!

There are a lot of questions here that are going to depend on your use-case. I think the community are better fit to answer most of them depending on their experience.

I will more focus on answering the technical questions.

> 2) Token Validation

In our documentation we have a guide for how to set up your own token validation: https://docs.serverpod.dev/concepts/authentication/custom-overrides
If you take an OATH2 approach, you could authenticated directly from the client towards Supabase and then implement the refresh token (JWT token) validation in Serverpod by overriding the `authenticationHandler`. This validation should only require you to fetch the public key of your Supabase authentication once. Then you can catch a potential 401 on the client and request a new refresh token from Supabase when this happens.

> 3) Data Integrity

I'm not completely sure what you mean here, but are you looking for transaction isolation levels? https://www.postgresql.org/docs/current/transaction-iso.html

> 5) MultipleDatabase:

Serverpod currently only supports connecting to a single database through the build in ORM. You are free to establish your own connection directly to a database if required, but then you won't draw the benefits from the typed ORM.
We hope to add support for multiple databases in the future.

**Comment by iampopal** [comment]
Thank you @SandPod 

If possible to get answers to the reminded questions will be very helpful
